File: LeetCode-July-Challenge/FindandReplacePattern.java
import java.util.*;
class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> result = new LinkedList<>();
        for (String word : words) {
            if (match(word, pattern)) {
                result.add(word);
            }
        }
        return result;
    }
    private boolean match(String s, String p) {
        Map<Character, Character> map1 = new HashMap<>();
        Map<Character, Character> map2 = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char c1 = s.charAt(i);
            char c2 = p.charAt(i);
            if (!map1.containsKey(c1)) {
                map1.put(c1, c2);
            }
            if (!map2.containsKey(c2)) {
                map2.put(c2, c1);
            }
            if (map1.get(c1) != c2 || map2.get(c2) != c1) {
                return false;
            }
        }
        return true;
    }
}
public class FindandReplacePattern {
    public static void main(String[] args) {
        String[] words = {"abc","deq","mee","aqq","dkd","ccc"};
        String pattern = "abb";
        Solution s = new Solution();
        List<String> res = s.findAndReplacePattern(words, pattern);
        for(int i = 0; i < res.size(); ++i){
            System.out.print(res.get(i) + " ");
        }
    }
}

File: LeetCode-July-Challenge/FindandReplacePattern.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> res;
        for(int i = 0; i < words.size(); ++i){
            map<char,char> mp1;
            map<char,char> mp2;
            int flag = 1;
            for(int j = 0; j < words[i].size(); ++j){
                auto it = mp1.find(pattern[j]);
                auto it2 = mp2.find(words[i][j]);
                if(it == mp1.end()){
                    mp1.insert({pattern[j], words[i][j]});
                } else{
                    if(it->second != words[i][j]){
                        flag = 0;
                        break;
                    }
                }
                if(it2 == mp2.end()){
                    mp2.insert({words[i][j], pattern[j]});
                } else{
                    if(it2->second != pattern[j]){
                        flag = 0;
                        break;
                    }
                }
            }
            if(flag) res.push_back(words[i]);
        }
        return res;
    }
};
int main(){
    vector<string> words = {"abc","deq","mee","aqq","dkd","ccc"};
    string pattern = "abb";
    Solution s;
    vector<string> res = s.findAndReplacePattern(words, pattern);
    for(int i = 0; i < res.size(); ++i){
        cout<<res[i]<<" ";
    }
    return 0;
}